### DISCUSSION TITLE
Speed up http post request

### DISCUSSION THREAD

**Discussion by pbelx** [vlang/v#19409] [Questions and Answers]
My net.http code is taking 41 seconds to get reply. Here is the code below.What could be wrong?

import net.http

fn main(){
	println("starting..")
	posturl := "http://127.0.0.1:2222/billing/user_signin/"
	post := '{"servedMSISDN": "2534444444","password": "dffdf4f7dd21d2293bd2d8500c507ad2c0414a156c6ea6f1d144bdc50cf244b82"}'
	h:= http.new_header_from_map({http.CommonHeader.content_type:'application/json'})
        println(h)
	conf :=http.FetchConfig{
	method:.post
	url:posturl,
	header:h,
	data:post
	user_agent:'volley-http'
	}
	resp := http.fetch(conf) !
	println(resp.body)
}

**Comment by JalonSolov** [comment]
Are you on Windows?  There are so many things in Windows that can slow you down... anti-virus, Windows Defender, etc., etc.

Since you didn't supply your server code, there is no way for anyone else to test, but I would be willing to bet it takes milliseconds or less on Linux for the same code...

**Reply by pbelx** [reply]
Running on debian 10.  Using curl command it takes around 9-10 seconds. But above code is always around 40-42 seconds.

**Reply by JalonSolov** [reply]
Are you certain your server code is responding fast enough?

You can try running with `-d trace_http_request` and `-d trace_http_response` to see what's happening.

You can also try using the `benchmark` module, and see if you can isolate the section(s) of code taking the most time.

**Reply by pbelx** [reply]
Have setup wireshark and can see that the json reply comes in earlier. 
The delay is on the program printing the response.

**Comment by joe-conigliaro** [comment]
Also it would be good to see how you are running it, is compilation included in that time?

**Reply by pbelx** [reply]
compilation not included. running a binary file.

**Comment by pbelx** [comment]
> My net.http code is taking 41 seconds to get reply. Here is the code below.What could be wrong?
> 
> import net.http
> 
> fn main(){ println("starting..") posturl := "http://127.0.0.1:2222/billing/user_signin/" post := '{"servedMSISDN": "2534444444","password": "dffdf4f7dd21d2293bd2d8500c507ad2c0414a156c6ea6f1d144bdc50cf244b82"}' h:= http.new_header_from_map({http.CommonHeader.content_type:'application/json'}) println(h) conf :=http.FetchConfig{ method:.post url:posturl, header:h, data:post user_agent:'volley-http' } resp := http.fetch(conf) ! println(resp.body) }

update using benchmark.

        import net.http
        import os
        import benchmark
        
        fn main(){
	        println("starting..")
	        mut b := benchmark.start()
	        posturl := "http://127.0.0.1:2222/billing/user_signin/"
	        post := '{"servedMSISDN": "252443434","password": "347f4f7dd21d2293bd2d8500c507ad2c0414a156c6ea6f1d144bdc50cf244b82"}'
        
        
	        
	        h:= http.new_header_from_map({http.CommonHeader.content_type:'application/json'})
	        b.measure('create header')
        
	        println(h)
	        b.measure('print header')
	        conf :=http.FetchConfig{
	        method:.post
	        url:posturl,
	        header:h,
	        data:post
	        user_agent:'volley-http'
	        }
	        b.measure('create conf')
	        resp := http.fetch(conf) !
	        b.measure('use fetch')
	        os.write_file("/tmp/js",resp.body) !
	        b.measure('write body')
	        println('done')
        

the results are.

starting..
 SPENT     0.021 ms in create header
Content-Type: application/json

 SPENT     0.009 ms in print header
 SPENT     0.001 ms in create conf
 SPENT 36102.196 ms in use fetch
 SPENT     0.312 ms in write body
done

real	0m36.109s
user	0m0.000s
sys	0m0.009s

**Reply by JalonSolov** [reply]
Now... can you add the same benchmarking code to your server?  If it takes 36 seconds to do the fetch... is it because the server doesn't respond for 36 seconds?

**Reply by pbelx** [reply]
Server responds within 9 seconds..http chunked json. running curl like this. 

    time curl -X POST http://127.0.0.1:2222/billing/user_signin/ -H 'Content-Type: application/json' -d '{"servedMSISDN": "25222222","password": "3d7f4f5dd21d2293bd2d8500c507ad2c0414a156c6ea6f1d144bdc50cf244b82"}'

        real	0m12.481s
        user 0m0.004s
        sys	0m0.012s

The V client is consistent at 36 seconds using fetch
 SPENT     0.010 ms in print header
 SPENT     0.001 ms in create conf
 SPENT 36071.013 ms in use fetch
 SPENT     0.291 ms in write body
done

real	0m36.106s
user	0m0.000s
sys	0m0.007s

**Reply by pbelx** [reply]
Here is a server reply format struct in Go Lang.

    type Reply2 struct {
	    UserBillingUsage struct {
		    UserName            string  `json:"userName"`
		    PostpaidCreditLimit any     `json:"postpaidCreditLimit"`
		    SubscriptionID      string  `json:"subscriptionId"`
		    ServedMSISDN        string  `json:"servedMSISDN"`
		    LastPayment         float64 `json:"lastPayment"`
		    PlanType            string  `json:"planType"`
		    DueAmount           any     `json:"dueAmount"`
		    CurrentUsage        float64 `json:"currentUsage"`
		    BillingCycle        any     `json:"billingCycle"`
		    FixedIP             any     `json:"fixedIp"`
		    VoiceUsage          struct {
			    TotalQuota            float64 `json:"totalQuota"`
			    CurrentUsage          float64 `json:"currentUsage"`
			    IsUnlimited           bool    `json:"isUnlimited"`
			    HasFallbackPlan       bool    `json:"hasFallbackPlan"`
			    CallUsageQuota        string  `json:"callUsageQuota"`
			    DataUsageQuota        any     `json:"dataUsageQuota"`
			    UsageHashMap          any     `json:"usageHashMap"`
			    VoiceAccumulatedUsage any     `json:"voiceAccumulatedUsage"`
			    DataAccumulatedUsage  any     `json:"dataAccumulatedUsage"`
			    SmsAccumulatedUsage   any     `json:"smsAccumulatedUsage"`
			    PerRecordUsage        []struct {
				    TariffType            string  `json:"tariffType"`
				    NTotalUnits           int     `json:"nTotalUnits"`
				    NLeftUnits            int     `json:"nLeftUnits"`
				    ServiceProvider       any     `json:"serviceProvider"`
				    RecToken              any     `json:"recToken"`
				    PlanID                string  `json:"planId"`
				    AdditionalQuotaID     any     `json:"additionalQuotaId"`
				    IsPlanAdditionalQuota bool    `json:"isPlanAdditionalQuota"`
				    NumeberOfUnusedCalls  any     `json:"numeberOfUnusedCalls"`
				    NumberOfCalls         any     `json:"numberOfCalls"`
				    MaxCallTime           any     `json:"maxCallTime"`
				    ServedMSISDN          string  `json:"servedMSISDN"`
				    StartDate             any     `json:"startDate"`
				    ValidityDate          any     `json:"validityDate"`
				    FinalUnitExp          any     `json:"finalUnitExp"`
				    UnitsType             string  `json:"unitsType"`
				    TariffUnitCost        any     `json:"tariffUnitCost"`
				    TariffUnitCurrency    int     `json:"tariffUnitCurrency"`
				    TariffIsActive        bool    `json:"tariffIsActive"`
				    LowbalanceNotified    bool    `json:"lowbalanceNotified"`
				    QuotaExceedTriggered  any     `json:"quotaExceedTriggered"`
				    ExpirationNotified    bool    `json:"expirationNotified"`
				    IsUnlimitedPlan       bool    `json:"isUnlimitedPlan"`
				    AboveQuotaSpeed       int     `json:"aboveQuotaSpeed"`
				    IsFallBackPlan        bool    `json:"isFallBackPlan"`
				    UseCreditLimit        bool    `json:"useCreditLimit"`
				    RoundoffType          any     `json:"roundoffType"`
				    IsMidNightExpiry      bool    `json:"isMidNightExpiry"`
				    DailyCap              int     `json:"dailyCap"`
				    DailyLeftCap          int     `json:"dailyLeftCap"`
				    MonthlyCap            int     `json:"monthlyCap"`
				    MonthlyLeftCap        int     `json:"monthlyLeftCap"`
				    DailyCapUnits         any     `json:"dailyCapUnits"`
				    DailyCapTimestamp     string  `json:"dailyCapTimestamp"`
				    MonthlyCapUnits       string  `json:"monthlyCapUnits"`
				    ParentSubscriptionID  any     `json:"parentSubscriptionId"`
				    IsSharedSubscription  bool    `json:"isSharedSubscription"`
				    FupQuotaUsed          int     `json:"fupQuotaUsed"`
				    RatingGroupID         any     `json:"ratingGroupId"`
				    CalledPartyLimit      any     `json:"calledPartyLimit"`
				    CalledParties         any     `json:"calledParties"`
				    NUsedUnits            int     `json:"nUsedUnits"`
				    IsDailyCap            any     `json:"isDailyCap"`
				    IsMonthlyCap          any     `json:"isMonthlyCap"`
				    IsCalledPartyLimit    any     `json:"isCalledPartyLimit"`
				    IsRegularTopup        any     `json:"isRegularTopup"`
				    FallBackUsedQuota     float64 `json:"fallBackUsedQuota"`
				    UnlimitedUsedQuota    float64 `json:"unlimitedUsedQuota"`
				    AmountChaged          float64 `json:"amountChaged"`
				    ExtraUsage            []any   `json:"extraUsage"`
				    PlanInfo              []struct {
					    First  string `json:"first"`
					    Second string `json:"second"`
				    } `json:"planInfo"`
				    RechargeTokens []string `json:"rechargeTokens"`
			    } `json:"perRecordUsage"`
			    HourlyUsageHistory  []any `json:"hourlyUsageHistory"`
			    DailyUsageHistory   []any `json:"dailyUsageHistory"`
			    MonthlyUsageHistory []any `json:"monthlyUsageHistory"`
		    } `json:"voiceUsage"`
		    SmsUsage struct {
			    TotalQuota            float64 `json:"totalQuota"`
			    CurrentUsage          float64 `json:"currentUsage"`
			    IsUnlimited           bool    `json:"isUnlimited"`
			    HasFallbackPlan       bool    `json:"hasFallbackPlan"`
			    CallUsageQuota        any     `json:"callUsageQuota"`
			    DataUsageQuota        any     `json:"dataUsageQuota"`
			    UsageHashMap          any     `json:"usageHashMap"`
			    VoiceAccumulatedUsage any     `json:"voiceAccumulatedUsage"`
			    DataAccumulatedUsage  any     `json:"dataAccumulatedUsage"`
			    SmsAccumulatedUsage   any     `json:"smsAccumulatedUsage"`
			    PerRecordUsage        []struct {
				    TariffType            string  `json:"tariffType"`
				    NTotalUnits           int     `json:"nTotalUnits"`
				    NLeftUnits            int     `json:"nLeftUnits"`
				    ServiceProvider       any     `json:"serviceProvider"`
				    RecToken              any     `json:"recToken"`
				    PlanID                string  `json:"planId"`
				    AdditionalQuotaID     any     `json:"additionalQuotaId"`
				    IsPlanAdditionalQuota bool    `json:"isPlanAdditionalQuota"`
				    NumeberOfUnusedCalls  any     `json:"numeberOfUnusedCalls"`
				    NumberOfCalls         any     `json:"numberOfCalls"`
				    MaxCallTime           any     `json:"maxCallTime"`
				    ServedMSISDN          string  `json:"servedMSISDN"`
				    StartDate             any     `json:"startDate"`
				    ValidityDate          any     `json:"validityDate"`
				    FinalUnitExp          any     `json:"finalUnitExp"`
				    UnitsType             string  `json:"unitsType"`
				    TariffUnitCost        any     `json:"tariffUnitCost"`
				    TariffUnitCurrency    int     `json:"tariffUnitCurrency"`
				    TariffIsActive        bool    `json:"tariffIsActive"`
				    LowbalanceNotified    bool    `json:"lowbalanceNotified"`
				    QuotaExceedTriggered  any     `json:"quotaExceedTriggered"`
				    ExpirationNotified    bool    `json:"expirationNotified"`
				    IsUnlimitedPlan       bool    `json:"isUnlimitedPlan"`
				    AboveQuotaSpeed       int     `json:"aboveQuotaSpeed"`
				    IsFallBackPlan        bool    `json:"isFallBackPlan"`
				    UseCreditLimit        bool    `json:"useCreditLimit"`
				    RoundoffType          any     `json:"roundoffType"`
				    IsMidNightExpiry      bool    `json:"isMidNightExpiry"`
				    DailyCap              int     `json:"dailyCap"`
				    DailyLeftCap          int     `json:"dailyLeftCap"`
				    MonthlyCap            int     `json:"monthlyCap"`
				    MonthlyLeftCap        int     `json:"monthlyLeftCap"`
				    DailyCapUnits         any     `json:"dailyCapUnits"`
				    DailyCapTimestamp     string  `json:"dailyCapTimestamp"`
				    MonthlyCapUnits       any     `json:"monthlyCapUnits"`
				    ParentSubscriptionID  any     `json:"parentSubscriptionId"`
				    IsSharedSubscription  bool    `json:"isSharedSubscription"`
				    FupQuotaUsed          int     `json:"fupQuotaUsed"`
				    RatingGroupID         any     `json:"ratingGroupId"`
				    CalledPartyLimit      any     `json:"calledPartyLimit"`
				    CalledParties         any     `json:"calledParties"`
				    NUsedUnits            int     `json:"nUsedUnits"`
				    IsDailyCap            any     `json:"isDailyCap"`
				    IsMonthlyCap          any     `json:"isMonthlyCap"`
				    IsCalledPartyLimit    any     `json:"isCalledPartyLimit"`
				    IsRegularTopup        any     `json:"isRegularTopup"`
				    FallBackUsedQuota     float64 `json:"fallBackUsedQuota"`
				    UnlimitedUsedQuota    float64 `json:"unlimitedUsedQuota"`
				    AmountChaged          float64 `json:"amountChaged"`
				    ExtraUsage            []any   `json:"extraUsage"`
				    PlanInfo              []struct {
					    First  string `json:"first"`
					    Second string `json:"second"`
				    } `json:"planInfo"`
				    RechargeTokens []string `json:"rechargeTokens"`
			    } `json:"perRecordUsage"`
			    HourlyUsageHistory  []any `json:"hourlyUsageHistory"`
			    DailyUsageHistory   []any `json:"dailyUsageHistory"`
			    MonthlyUsageHistory []any `json:"monthlyUsageHistory"`
		    } `json:"smsUsage"`
		    DataUsage struct {
			    TotalQuota            float64 `json:"totalQuota"`
			    CurrentUsage          float64 `json:"currentUsage"`
			    IsUnlimited           bool    `json:"isUnlimited"`
			    HasFallbackPlan       bool    `json:"hasFallbackPlan"`
			    CallUsageQuota        any     `json:"callUsageQuota"`
			    DataUsageQuota        string  `json:"dataUsageQuota"`
			    UsageHashMap          any     `json:"usageHashMap"`
			    VoiceAccumulatedUsage any     `json:"voiceAccumulatedUsage"`
			    DataAccumulatedUsage  any     `json:"dataAccumulatedUsage"`
			    SmsAccumulatedUsage   any     `json:"smsAccumulatedUsage"`
			    PerRecordUsage        []struct {
				    TariffType            string  `json:"tariffType"`
				    NTotalUnits           int64   `json:"nTotalUnits"`
				    NLeftUnits            int     `json:"nLeftUnits"`
				    ServiceProvider       any     `json:"serviceProvider"`
				    RecToken              any     `json:"recToken"`
				    PlanID                string  `json:"planId"`
				    AdditionalQuotaID     any     `json:"additionalQuotaId"`
				    IsPlanAdditionalQuota bool    `json:"isPlanAdditionalQuota"`
				    NumeberOfUnusedCalls  any     `json:"numeberOfUnusedCalls"`
				    NumberOfCalls         any     `json:"numberOfCalls"`
				    MaxCallTime           any     `json:"maxCallTime"`
				    ServedMSISDN          string  `json:"servedMSISDN"`
				    StartDate             any     `json:"startDate"`
				    ValidityDate          any     `json:"validityDate"`
				    FinalUnitExp          any     `json:"finalUnitExp"`
				    UnitsType             string  `json:"unitsType"`
				    TariffUnitCost        any     `json:"tariffUnitCost"`
				    TariffUnitCurrency    int     `json:"tariffUnitCurrency"`
				    TariffIsActive        bool    `json:"tariffIsActive"`
				    LowbalanceNotified    bool    `json:"lowbalanceNotified"`
				    QuotaExceedTriggered  any     `json:"quotaExceedTriggered"`
				    ExpirationNotified    bool    `json:"expirationNotified"`
				    IsUnlimitedPlan       bool    `json:"isUnlimitedPlan"`
				    AboveQuotaSpeed       int     `json:"aboveQuotaSpeed"`
				    IsFallBackPlan        bool    `json:"isFallBackPlan"`
				    UseCreditLimit        bool    `json:"useCreditLimit"`
				    RoundoffType          any     `json:"roundoffType"`
				    IsMidNightExpiry      bool    `json:"isMidNightExpiry"`
				    DailyCap              int     `json:"dailyCap"`
				    DailyLeftCap          int     `json:"dailyLeftCap"`
				    MonthlyCap            int     `json:"monthlyCap"`
				    MonthlyLeftCap        int     `json:"monthlyLeftCap"`
				    DailyCapUnits         any     `json:"dailyCapUnits"`
				    DailyCapTimestamp     string  `json:"dailyCapTimestamp"`
				    MonthlyCapUnits       any     `json:"monthlyCapUnits"`
				    ParentSubscriptionID  any     `json:"parentSubscriptionId"`
				    IsSharedSubscription  bool    `json:"isSharedSubscription"`
				    FupQuotaUsed          int     `json:"fupQuotaUsed"`
				    RatingGroupID         any     `json:"ratingGroupId"`
				    CalledPartyLimit      any     `json:"calledPartyLimit"`
				    CalledParties         any     `json:"calledParties"`
				    NUsedUnits            int64   `json:"nUsedUnits"`
				    IsDailyCap            any     `json:"isDailyCap"`
				    IsMonthlyCap          any     `json:"isMonthlyCap"`
				    IsCalledPartyLimit    any     `json:"isCalledPartyLimit"`
				    IsRegularTopup        any     `json:"isRegularTopup"`
				    FallBackUsedQuota     float64 `json:"fallBackUsedQuota"`
				    UnlimitedUsedQuota    float64 `json:"unlimitedUsedQuota"`
				    AmountChaged          float64 `json:"amountChaged"`
				    ExtraUsage            []any   `json:"extraUsage"`
				    PlanInfo              []struct {
					    First  string `json:"first"`
					    Second string `json:"second"`
				    } `json:"planInfo"`
				    RechargeTokens []string `json:"rechargeTokens"`
			    } `json:"perRecordUsage"`
			    HourlyUsageHistory  []any `json:"hourlyUsageHistory"`
			    DailyUsageHistory   []any `json:"dailyUsageHistory"`
			    MonthlyUsageHistory []any `json:"monthlyUsageHistory"`
		    } `json:"dataUsage"`
		    LatestBill          float64 `json:"latestBill"`
		    DailyCapQuota       int     `json:"dailyCapQuota"`
		    DailyCapUsage       int     `json:"dailyCapUsage"`
		    MonthlyCapQuota     int     `json:"monthlyCapQuota"`
		    MonthlyCapUsage     int     `json:"monthlyCapUsage"`
		    QuotaUsage          any     `json:"quotaUsage"`
		    TotalQuota          float64 `json:"totalQuota"`
		    HourlyUsageHistory  []any   `json:"hourlyUsageHistory"`
		    DailyUsageHistory   []any   `json:"dailyUsageHistory"`
		    MonthlyUsageHistory []any   `json:"monthlyUsageHistory"`
		    PlanName            string  `json:"planName"`
		    ValidUntill         string  `json:"validUntill"`
	    } `json:"userBillingUsage"`
	    Status string `json:"status"`
    }

**Reply by joe-conigliaro** [reply]
It would be ideal if you could setup a server/client script to demonstrate the issue.
Or at least provide the response from your server, my guess is this has something to do with chunking.
